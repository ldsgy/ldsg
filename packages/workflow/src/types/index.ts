export * from "./get-workflow-node-type-info";
export * from "./specific-resource-settings";
