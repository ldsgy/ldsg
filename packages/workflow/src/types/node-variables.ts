export type NodeId = string;

export type NodeVariables = any;

export type NodeIdToVariablesMap = Map<NodeId, NodeVariables>;
