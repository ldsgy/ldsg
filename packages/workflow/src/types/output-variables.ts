export type NodeId = string;

export type NodeOutputVariables = any;

export type NodeIdToOutputVariablesMap = Record<NodeId, NodeOutputVariables>;
