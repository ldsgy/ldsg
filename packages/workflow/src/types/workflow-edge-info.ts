export interface WorkflowEdgeInfo {
  /**
   * 起点工作流节点资源ID
   */
  sourceWorkflowNodeResourceId: string;

  /**
   * 起点工作流节点端口ID
   */
  sourceWorkflowNodePortId?: string;

  /**
   * 目标工作流节点资源ID
   */
  targetWorkflowNodeResourceId: string;

  /**
   * 目标工作流节点端口ID
   */
  targetWorkflowNodePortId?: string;
}

export interface GetWorkflowEdgeInfoRes {
  /**
   * Workflow Edge Info
   */
  workflowEdgeInfo: WorkflowEdgeInfo;
}

export type GetWorkflowEdgeInfo = () => GetWorkflowEdgeInfoRes;
