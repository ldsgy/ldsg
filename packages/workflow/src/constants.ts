import { ROOT_RESOURCE_KIND } from "@ldsg/constants";
import { HandlerSpecificResourceSettings } from "@ldsg/handler";
import { ResourceDefinitionSpecificResourceSettings } from "@ldsg/resource-definition";
import { GeneralResourceSettings } from "@ldsg/types";

export const HANDLER_GENERAL_RESOURCE_SETTINGS: GeneralResourceSettings = {
  title: "工作流类型资源处理程序",
  description: "",
};

/**
 * Handler Resource Settings
 * 此类型资源的子级处理程序资源的配置
 * 如无或为 undefined，则说明无需子级处理程序。
 */
export const HANDLER_SPECIFIC_RESOURCE_SETTINGS: HandlerSpecificResourceSettings =
  {
    code: `export * from "@ldsg/workflow";`,
    dependencies: [
      {
        name: "@ldsg/workflow",
      },
    ],
  };

export const RESOURCE_DEFINITION_GENERAL_RESOURCE_SETTINGS: GeneralResourceSettings =
  {
    title: "工作流类型资源定义",
    description: "",
  };

export const RESOURCE_DEFINITION_SPECIFIC_RESOURCE_SETTINGS: ResourceDefinitionSpecificResourceSettings =
  {
    kind: "workflow_node_type",
    parentKind: ROOT_RESOURCE_KIND,
  };

/**
 * Specific Workflow Node Type Id Map
 */
export const SPECIFIC_WORKFLOW_NODE_TYPE_TO_WORKFLOW_NODE_TYPE_RESOURCE_ID_MAP =
  {
    start: "workflow_node_type_start",
    end: "workflow_node_type_end",
  } as const;
