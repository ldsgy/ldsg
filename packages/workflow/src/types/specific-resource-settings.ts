import { SpecificResourceSettings } from "@ldsg/types";

export interface WorkflowSpecificResourceSettings
  extends SpecificResourceSettings {}
