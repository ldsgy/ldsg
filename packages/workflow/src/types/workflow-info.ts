import { GetEndNodeOutputVariablesRes } from "../utils";
import { NodeVariables } from "./node-variables";

export interface WorkflowExecuteParams {
  startNodeInputVariables: NodeVariables;
}

export interface WorkflowExecuteRes {
  endNodeOutputVariables: NodeVariables;
}

export type WorkflowExecute = (
  params: WorkflowExecuteParams
) => Promise<GetEndNodeOutputVariablesRes>;

export interface WorkflowInfo {
  execute: WorkflowExecute;
}
