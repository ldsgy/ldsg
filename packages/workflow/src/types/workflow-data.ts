import { WorkflowEdgeInfo } from "./workflow-edge-info";
import { WorkflowNodeInfo } from "./workflow-node-info";

export interface WorkflowData {
  edges: WorkflowEdgeInfo[];
  nodes: WorkflowNodeInfo[];
}
