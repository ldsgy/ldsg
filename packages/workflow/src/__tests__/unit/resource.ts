import { SPECIFIC_WORKFLOW_NODE_TYPE_TO_WORKFLOW_NODE_TYPE_RESOURCE_ID_MAP } from "../../constants";
import {
  GetWorkflowData,
  GetWorkflowDataRes,
  WorkflowResource,
} from "../../resource";
import { WorkflowData, WorkflowNodeExecuter } from "../../types";

class AWorkflowResource extends WorkflowResource {
  getWorkflowData: GetWorkflowData = () => {
    const workflowData: WorkflowData = {
      edges: [
        {
          sourceWorkflowNodeResourceId: "1",
          targetWorkflowNodeResourceId: "2",
        },
        {
          sourceWorkflowNodeResourceId: "2",
          targetWorkflowNodeResourceId: "3",
        },
        {
          sourceWorkflowNodeResourceId: "3",
          targetWorkflowNodeResourceId: "4",
        },
        {
          sourceWorkflowNodeResourceId: "4",
          targetWorkflowNodeResourceId: "5",
        },
      ],
      nodes: [
        {
          id: "1",
          workflowNodeTypeResourceId:
            SPECIFIC_WORKFLOW_NODE_TYPE_TO_WORKFLOW_NODE_TYPE_RESOURCE_ID_MAP.start,
          Executer: class a extends WorkflowNodeExecuter {},
        },
        {
          id: "2",
          workflowNodeTypeResourceId: "",
          Executer: class a extends WorkflowNodeExecuter {
            execute = () => {
              const { setOutputVariables } = this;

              setOutputVariables({
                outputVariables: "abc",
              });
            };
          },
        },
        {
          id: "3",
          workflowNodeTypeResourceId: "",
          Executer: class a extends WorkflowNodeExecuter {
            execute = () => {
              const { setOutputVariables } = this;

              setOutputVariables({
                outputVariables: "def",
              });
            };
          },
        },
        {
          id: "4",
          workflowNodeTypeResourceId: "",
          Executer: class a extends WorkflowNodeExecuter {
            execute = () => {
              const { setOutputVariables } = this;

              setOutputVariables({
                outputVariables: "ghi",
              });
            };
          },
        },
        {
          id: "5",
          workflowNodeTypeResourceId:
            SPECIFIC_WORKFLOW_NODE_TYPE_TO_WORKFLOW_NODE_TYPE_RESOURCE_ID_MAP.end,
          Executer: class a extends WorkflowNodeExecuter {},
        },
      ],
    };

    const res: GetWorkflowDataRes = {
      workflowData,
    };

    return res;
  };
}

test("WorkflowResource execute", async () => {
  const aWorkflowResource = new AWorkflowResource({
    id: "test",
    kind: "workflow",
    parentId: "root",
    settings: {
      title: "",
      description: "",
    },
  });

  await aWorkflowResource.execute({
    startNodeOutputVariables: {
      hello: "world",
    },
  });

  expect(aWorkflowResource.nodeIdToOutputVariablesMap).toMatchSnapshot();
});
