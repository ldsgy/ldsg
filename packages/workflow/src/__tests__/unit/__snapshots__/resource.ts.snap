// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkflowResource execute 1`] = `"jkl"`;
